libs/execution_context/ebpf_maps.c

```diff
@@ -1209,6 +1337,74 @@ _delete_lru_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
 static ebpf_result_t
 _delete_hash_map_entry(_Inout_ ebpf_core_map_t* map, _In_ const uint8_t* key);
 
+/**
+ * @brief Walk the cold lists, removing secondary keys and finding the oldest primary key.
+ *
+ * @param[in] lru_map
+ * @return Either the oldest primary key or NULL if no primary keys are present.
+ */
+static ebpf_lru_entry_t*
+_reap_lru_cold_lists(ebpf_core_lru_map_t* lru_map)
+{
+    ebpf_lru_entry_t* oldest_entry = NULL;
+    uint64_t oldest_timestamp = UINT64_MAX;
+    bool non_empty_partition_found = true;
+
+    // Scan all partitions until we find a non-empty cold list or we have an oldest entry.
+    while (!oldest_entry && non_empty_partition_found) {
+        non_empty_partition_found = false;
+        // Scan each partition, removing secondary entries with higher timestamps until we find the oldest entry.
+        for (size_t partition = 0; partition < lru_map->partition_count; partition++) {
+            // If this cold list is empty, skip it.
+            if (ebpf_list_is_empty(&lru_map->partitions[partition].cold_list)) {
+                continue;
+            }
+
+            // Found a non-empty partition.
+            non_empty_partition_found = true;
+
+            // Lock the partition.
+            ebpf_lock_state_t state = ebpf_lock_lock(&lru_map->partitions[partition].lock);
+
+            // Check again after acquiring the lock.
+            // If the cold list is empty, skip it.
+            if (ebpf_list_is_empty(&lru_map->partitions[partition].cold_list)) {
+                ebpf_lock_unlock(&lru_map->partitions[partition].lock, state);
+                continue;
+            }
+
+            // Compute the start of the entry from the cold list entry.
+            ebpf_list_entry_t* list_entry = lru_map->partitions[partition].cold_list.Flink - partition;
+            ebpf_lru_entry_t* entry = (ebpf_lru_entry_t*)list_entry;
+
+            // The effective age of an entry is the maximum of the last used time of all partitions.
+            // Compute the highest timestamp for this entry.
```

Comment:
Just a comment: depending on the way / order the entries are accessed on each CPU, this whole loop may run for a very long time before we find an entry to delete. So the worst-case performance may be very bad.

Reply:
Agreed. I would recommend we keep an eye on this and if there are complaints about LRU replacement time we revisit this.